### PR TITLE
[JENKINS-66470] Register view-related permission groups and permissions prior to the execution of JCasC

### PR DESCRIPTION
**Pull request by basil**
See [JENKINS-66470](https://issues.jenkins.io/browse/JENKINS-66470).

Prior to the removal of `TreeView` in #5603, the registration of view-related permission groups and permissions prior to the execution of JCasC (i.e., before `InitMilestone.SYSTEM_CONFIG_LOADED`) worked more or less by accident: by virtue of `TreeView` happening to have a static method annotated with `@Extension`, which happened to be executed very early (in my tests, because of Git plugin's `GitSCM#onLoaded` method), which happened to load the `TreeView` class, which happened to load the `View` class, which happened to run its static initializers, which happened to register the view-related permission groups and permissions. With the removal of `TreeView`, this is no longer happening, and view-related permission groups and permissions are not being registered prior to `InitMilestone.SYSTEM_CONFIG_LOADED`, causing problems with JCasC configurations that reference these permission groups and permissions.

The long-term fix is described in [JENKINS-17200](https://issues.jenkins.io/browse/JENKINS-17200). This PR restores the status quo by ensuring that view-related permission groups and permissions are registered prior to `InitMilestone.SYSTEM_CONFIG_LOADED` (after which JCasC needs them) by adding a new static method in the `View` class. This new static method is placed after the static construction of the permission group and permission instances, during which these instances are registered. By annotating the method with `@Initializer(before = InitMilestone.SYSTEM_CONFIG_LOADED)`, we ensure that the registration takes place before JCasC starts consuming these permissions. Thus the status quo is restored.

### Testing done

Reproduced the problem as described in jenkinsci/configuration-as-code-plugin#1638; could no longer do so after these changes.

### Proposed changelog entries

* Allow Jenkins to start when the JCasC configuration defines view-related permissions (regression in 2.302).

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by timja — APPROVED** [review]
Nice!

**Review by res0nance — APPROVED** [review]
Also 👍 for labelling it `lts-candidate` we should definitely backport this

**Comment by basil** [comment]
Reviewers might be tempted to think that this PR causes a new test failure in `hudson.model.ViewTest.testGetComputers`; however, that is not the case. See, for example, [this build of unrelated PR #5722](https://ci.jenkins.io/job/Core/job/jenkins/job/PR-5722/3/), which also shows the same failure. That test creates 5 agents, which seems to be above the capacity of whatever environment the test is currently running in. Maintaining a compute environment that is sufficient for the needs of the core test suite is the domain of the infrastructure team and is unrelated to this PR.

File: core/src/main/java/hudson/model/View.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/View.java:1323]:
```suggestion
    @Initializer(before = InitMilestone.SYSTEM_CONFIG_LOADED)
    @Restricted(DoNotUse.class)
```

**Review by daniel-beck — APPROVED** [review]
Looks like a suitable workaround, thanks for taking care of this!

File: core/src/main/java/hudson/model/View.java
Comment (jglick) [inline core/src/main/java/hudson/model/View.java:1329]: Or you could be explicit like

```java
Objects.hash(CREATE, DELETE, CONFIGURE, READ);
```

**Review by jglick — COMMENTED** [review]
Is there some reason this initializer needs to be core? Does the same method in `configuration-as-code` work, or is that too late (has to run before plugin class loaders are even created)?

**Comment by basil** [comment]
> Is there some reason this initializer needs to be core? Does the same method in `configuration-as-code` work, or is that too late (has to run before plugin class loaders are even created)?

Not sure.

File: core/src/main/java/hudson/model/View.java
Comment (basil) [reply core/src/main/java/hudson/model/View.java:1329]: Could indeed.

**Comment by timja** [comment]
> Is there some reason this initializer needs to be core? Does the same method in configuration-as-code work, or is that too late (has to run before plugin class loaders are even created)?

It's not directly related to jcasc as such, each authorization plugin has written a custom model for jcasc.

The authorization plugins are generic and know nothing about registered permissions, so I don't think it fits there.

The place defining the permission is responsible for it being loaded (until the linked RFE is implemented that makes it declarative)

File: core/src/main/java/hudson/model/View.java
Comment (basil) [reply core/src/main/java/hudson/model/View.java:1324]: That'll teach me to accept "suggestions" on GitHub.

Comment (jglick) [reply core/src/main/java/hudson/model/View.java:1324]: caveat acceptor

**Comment by jglick** [comment]
> The place defining the permission is responsible for it being loaded

Is this to say that `Permission`s in non-`View` groups were already being eagerly loaded in time due to other `@Initializer`s or `@Extension`s?

**Comment by timja** [comment]
Yes, never seen this issue before this happened

File: core/src/main/java/hudson/model/View.java
Comment (daniel-beck) [reply core/src/main/java/hudson/model/View.java:1324]: I checked when I suggested and both classes are imported?

**Comment by daniel-beck** [comment]
> Yes, never seen this issue before this happened

I wonder whether that applies to the (probably barely used?) SCM/Tag permission that JENKINS-17200 was filed for. I've seen it not loaded in Jenkins instances without jobs before. That was a few years ago though, so may have changed in the mean time.

File: core/src/main/java/hudson/model/View.java
Comment (basil) [reply core/src/main/java/hudson/model/View.java:1324]: You're not missing anything. No imports are needed, hence my joke about accepting suggestions on GitHub (the point of the joke being that even when such suggestions don't lead to compilation errors, they do often lead to unnecessary review feedback :smile:).

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: core/src/main/java/hudson/model/View.java
Comment (timja) [reply core/src/main/java/hudson/model/View.java:1324]: Import ordering lol, they're in the exact same package yet miles away in the imports.

I just saw `NoExternalUse` and `DoNotUse` wasn't on my screen.

Comment (basil) [reply core/src/main/java/hudson/model/View.java:1324]: > Import ordering lol, they're in the exact same package yet miles away in the imports.

I've been debating whether or not to sort all the imports in core. It would take me five minutes of course, but not sure how receptive others would be. :smile:

Comment (timja) [reply core/src/main/java/hudson/model/View.java:1324]: me too, was just thinking of doing it via spotless

Comment (basil) [reply core/src/main/java/hudson/model/View.java:1324]: => #5732

**Comment by BerndFarkaDyna** [comment]
+1 from my side, we are blocked with upgrading due to this issue

**Pull request merged**
